java/arcs/android/devtools/DevWebSocketFactory.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.devtools
+
+import arcs.core.util.TaggedLog
+import fi.iki.elonen.NanoHTTPD
+import fi.iki.elonen.NanoWSD
+import fi.iki.elonen.NanoWSD.WebSocketFrame.CloseCode
+import java.io.IOException
+
+/**
+ * An extension of [NanoWSD] for devtools to connect Arcs to a remote device for debugging.
+ */
+class DevWebSocketFactory {
+
+    fun getDevWebSocket() = devWebSocket
+
+    companion object devWebSocket : NanoWSD("localhost", 33317) {
+        private var wsdSocket: WsdSocket? = null
+        private var started = false
+        private val log = TaggedLog { "DevWebSocket" }
+
+        /**
+         * Send a string to the client.
+         */
+        fun send(msg: String) {
+            if (wsdSocket?.open ?: false) {
+                wsdSocket?.send(msg)
+            } else {
+                log.debug { "WebSocket Closed, can't send message [message=$msg]." }
+            }
+        }
+
+        override fun openWebSocket(ihttpSession: NanoHTTPD.IHTTPSession?): WebSocket {
+            wsdSocket = WsdSocket(ihttpSession, log)
+            return wsdSocket!!
+        }
+
+        fun startIfNeeded() {
+            if (!started) {
+                super.start()
+                started = true
+            }
+        }
```

Comment:
If the web server is a singleton and is only ever created/launched once when the DevToolsService's `onCreate` or something is triggered, this may not be necessary.